New datasources: kubernetes_namespace & kubernetes_all_namespaces (#701)
terraform-providers/terraform-provider-kubernetes: website/docs/d/all_namespaces.html.markdown

---
layout: "kubernetes"
page_title: "Kubernetes: kubernetes_all_namespaces"
sidebar_current: "docs-kubernetes-data-source-all-namespaces"
description: |-
  Lists all namespaces within a cluster.
---

# kubernetes_all_namespaces

This data source provides a mechanism for listing the names of all available namespaces in a Kubernetes cluster.

> It can be used check for existence of a specific namespaces or to apply another resource to all or a subset of existing namespaces in a cluster.

Reviewer: ```suggestion
It can be used to check for existence of a specific namespaces or to apply another resource to all or a subset of existing namespaces in a cluster.
```

> In Kubernetes, namespaces provide a scope for names and are intended as a way to divide cluster resources between multiple users.

Reviewer: What about using a description closer to the official one which is a bit more explicit?

```suggestion
In Kubernetes, namespaces are intended for use in environments with many users spread across multiple teams, or projects. For clusters with a few to tens of users, you should not need to create or think about namespaces at all. Start using namespaces when you need the features they provide. 

Namespaces provide a scope for names. Names of resources need to be unique within a namespace, but not across namespaces. Namespaces can not be nested inside one another and each Kubernetes resource can only be in one namespace.

Namespaces are a way to divide cluster resources between multiple users (via resource quota).
```

From https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#when-to-use-multiple-namespaces

## Example Usage

```hcl
data "kubernetes_all_namespaces" "allns" {}

output "all-ns" {
  value = data.kubernetes_all_namespaces.allns.namespaces
}

output "ns-present" {
  value = contains(data.kubernetes_all_namespaces.allns.namespaces, "kube-system")
}

> ```

Reviewer: Datasources usually have an attribute reference section:
````suggestion
```

## Attribute Reference

The following attributes are exported:

* `namespaces` - A list of existing namespace names.
````

See:
- https://github.com/terraform-providers/terraform-provider-kubernetes/blame/v1.10.0/website/docs/d/secret.html.markdown#L51-L54
- https://github.com/terraform-providers/terraform-provider-google/blame/v2.20.0/website/docs/d/datasource_compute_region_instance_group.html.markdown#L58-L69
- https://github.com/terraform-providers/terraform-provider-null/blame/v2.1.2/website/docs/data_source.html.markdown#L55-L64
